src/GNDStk/Tree/src/ctor.hpp

```diff
@@ -0,0 +1,136 @@
+
+// -----------------------------------------------------------------------------
+// Constructors
+// -----------------------------------------------------------------------------
+
+// -----------------------------------------------------------------------------
+// Basics
+// -----------------------------------------------------------------------------
+
+// default, move
+Tree() { }
+Tree(Tree &&) = default;
+
+// copy
+Tree(const Tree &from)
+{
+   convert(from,*this);
+}
+
+// templated "copy"
+template<
+   template<class...> class METADATA_CONTAINER_FROM,
+   template<class...> class CHILDREN_CONTAINER_FROM
+>
```

Comment:
Ah, but they're actually different, and quite intentionally so. Note that the enclosing Tree's template arguments are:

```
   METADATA_CONTAINER,
   CHILDREN_CONTAINER
```

while these are:

```
   METADATA_CONTAINER_FROM,
   CHILDREN_CONTAINER_FROM
```

i.e. with a _FROM suffix. Tree wants a container type for its metadata, and a container type for its children. Two different Tree objects might have the same container types, or completely different ones. Regardless, we should be able to make one from the other.

Analogy: Say we have vector<T>, just like the C++ STL has. It's perfectly reasonable - more than reasonable, actually - to have these two assignment operators:

```
   vector<T> = vector<T>  // copy assignment
   vector<T> = vector<X>  // where X is not the same as T
```

and likewise for, say, vector's constructors. This is really all I'm doing for Tree.

Reply:
I'll comment on this in another place, where you asked the same question.

Reply:
The other place doesn't show a reply box, so I'll reply here...

Generally, I'll put in this sort of ability to be generic if any of the following apply:

   * It's needed. (Of course.)

   * It's probably not needed, but is trivial to support.

   * It may prove to be useful, and isn't much extra effort to support, even if perhaps it's not completely trivial.

The last of these, I believe, applies here. I don't think that there's really very much added code due to this.

One consideration is that I do make the metadata and children containers publicly accessible to users. That way, with no extra code effort by us, the user has access to whatever functionality the container in question allows. (Then we just need to provide additional functionality that arguably adds real value, to this specific library, above and beyond what the container provides.)

A particular user may be more familiar with one container type than another. Given a choice, users can select whatever container type they're most effective with, instead of being required to use a particular one that we foist on them. (And we have a default, so that an less-knowledgeable user just uses the default.)

That said, there are really only three containers that I *know* work with the code in its existing form. This is due to particular functions that I assume the given container supports. I've been meaning to look at that sometime, and see if extending the number of allowable C++ standard containers can be increased without adding much complexity. That's lower down on my to-do list precisely because, while I think the added code to support a templated container is useful for the reasons I'm discussing, I don't see it as overwhelmingly important.

I've also found that genericity, early on, can make it much easier later, when the code is out in the wild, to collect informal statistics about its performance (speed and memory). Armed with information about performance of real user codes, we can easily change container types, or have users do so, and get a feel for what might work better than our original choice. One could certainly go in and just change a hardcoded container, recompile, get new performance data, change again, recompile, get more performance data, and so forth. (Easier if the actual container is spelled out in only one place, harder if it's hardcoded in several places.) Invariably, that's more work than being able to try other options quickly with a template parameter change. (And, with the template parameter, one could make a single code that instantiates several possibilities and measures their performance all in one fell swoop. At least one of my test codes tries lots of different combinations, although it doesn't do anything to measure performance.)

Basically, the above was my reasoning for making the container types be template arguments.